source/common/formatter/substitution_formatter.cc

```diff
@@ -183,8 +183,13 @@ StructFormatter::FormatBuilder::toFormatMapValue(const ProtobufWkt::Struct& stru
       output->emplace(pair.first, toFormatListValue(pair.second.list_value()));
       break;
 
+    case ProtobufWkt::Value::kNumberValue:
+      output->emplace(pair.first,
+                      toFormatStringValue(std::to_string(uint32_t(pair.second.number_value()))));
+      break;
```

Comment:
I don't think it's correct to force the value to be a uint32_t. Number Values are [stored as signed doubles](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/struct.proto#L68).

I think it would be better to provide a `class PlainNumberFormatter : FormatterProvider` (similar to the existing `PlainStringFormater`). That requires implement `format` (which should emit a string using with the same semantics as the `%g` printf specifier) and `formatValue` (which can just return the number Value).

You can then provide a `toFormatNumberValue` method in this class and use it to create a PlainNumberFormatter which the concrete structured output class can emit in the appropriate format. I think that's limited to JSON right now, which uses the same type for numeric output.

Reply:
Done.